test/integration/ws/common.dart

```diff
@@ -325,31 +329,133 @@ void runCommonWebSocketIntegrationTests(
     sub.pause();
-    await new Future.delayed(new Duration(milliseconds: 500));
+    await new Future.delayed(new Duration(milliseconds: 100));
     webSocket.add('3');
-    await new Future.delayed(new Duration(milliseconds: 500));
+    await new Future.delayed(new Duration(milliseconds: 100));
 
     // resume the subscription, messages should be recorded again
     sub.resume();
-    await new Future.delayed(new Duration(milliseconds: 500));
+    await new Future.delayed(new Duration(milliseconds: 100));
     webSocket.add('4');
-    await new Future.delayed(new Duration(milliseconds: 500));
+    await new Future.delayed(new Duration(milliseconds: 100));
 
     expect(messages, orderedEquals(['2', '4']));
     await webSocket.close();
   });
 
+  test('should support calling pause() with a resume signal', () async {
+    var webSocket = await connect(echoUri);
+    var messages = [];
+
+    // setup a subscription, messages should be recorded
+    var sub = webSocket.listen((data) {
+      messages.add(data);
+    });
+    webSocket.add('1');
+    await new Future.delayed(new Duration(milliseconds: 100));
+
+    // pause the subscription, messages should be discarded until the resume
+    // signal future resolves.
+    var c = new Completer();
+    sub.pause(c.future);
+    await new Future.delayed(new Duration(milliseconds: 100));
+    webSocket.add('2');
+    await new Future.delayed(new Duration(milliseconds: 100));
+
+    // resume the subscription, messages should be recorded again
+    c.complete();
+    await new Future.delayed(new Duration(milliseconds: 100));
+    webSocket.add('3');
+    await new Future.delayed(new Duration(milliseconds: 100));
+
+    expect(messages, orderedEquals(['1', '3']));
+    await webSocket.close();
+  });
+
+  test(
+      'should support calling pause() with a resume signal even if it resolves with an error',
+      () async {
+    var webSocket = await connect(echoUri);
+    var messages = [];
+
+    // setup a subscription, messages should be recorded
+    var sub = webSocket.listen((data) {
+      messages.add(data);
+    });
+    webSocket.add('1');
+    await new Future.delayed(new Duration(milliseconds: 100));
+
+    // pause the subscription, messages should be discarded until the resume
+    // signal future resolves.
+    var c = new Completer();
+    sub.pause(c.future);
+    await new Future.delayed(new Duration(milliseconds: 100));
+    webSocket.add('2');
+    await new Future.delayed(new Duration(milliseconds: 100));
+
+    // resume the subscription, messages should be recorded again
+    c.completeError(new Exception('Ignore. This error is expected.'));
+    await new Future.delayed(new Duration(milliseconds: 100));
+    webSocket.add('3');
+    await new Future.delayed(new Duration(milliseconds: 100));
+
+    expect(messages, orderedEquals(['1', '3']));
+    await webSocket.close();
+  }, skip: 'Can\'t test without the exception causing the test to fail.');
```

Comment:
Yeah unfortunately that only works when the error is thrown synchronously or when a future resolves with an error. In this case, the error is thrown asynchronously and can't be caught like that.